docs/root/configuration/http_conn_man/headers.rst

```diff
@@ -166,17 +167,117 @@ accurate. The source IP address of the immediate downstream node's connection to
 trusted. XFF *sometimes* can be trusted. Malicious clients can forge XFF, but the last
 address in XFF can be trusted if it was put there by a trusted proxy.
 
-Envoy's rules for determining the trusted client address are:
+Envoy's default rules for determining the trusted client address (*before* appending anything
+to XFF) are:
 
 * If *use_remote_address* is false and an XFF containing at least one IP address is
   present in the request, the trusted client address is the *last* (rightmost) IP address in XFF.
 * Otherwise, the trusted client address is the source IP address of the immediate downstream
   node's connection to Envoy.
 
+In an environment where there are one or more trusted proxies in front of an edge
+Envoy instance, the *xff_num_trusted_hops* configuration option can be used to trust
+additional addresses from XFF:
+
+* If *use_remote_address* is false and *xff_num_trusted_hops* is set to a value *N* that is
+  greater than zero, the trusted client address is the (N+1)th address from the right end
+  of XFF. (If the XFF contains fewer than N+1 addresses, Envoy falls back to using the
+  immediate downstream connection's source address as trusted client address.)
+* If *use_remote_address* is true and *xff_num_trusted_hops* is set to a value *N* that is
+  greater than zero, the trusted client address is the Nth address from the right end
+  of XFF. (If the XFF contains fewer than N addresses, Envoy falls back to using the
+  immediate downstream connection's source address as trusted client address.)
+
```

Comment:
So first off, thank you for this amazing documentation, it makes it really clear what is going on over in Envoy #2559

The goal here is to preserve existing Envoy behavior while allowing configuring xff_num_trusted_hops, right?  The one thing I'm still struggling with is the use of the combinations and permutations.  If we were designing this from scratch, would there be a gain in having the two separate config options, or would we only want one, where we always considered the first hop to be the direct connected host?

Reply:
Based on the discussion in Envoy #[2559](https://github.com/envoyproxy/envoy/pull/2559), I created a follow-up issue [2590](https://github.com/envoyproxy/envoy/issues/2590) to simplify the configuration.